src/main/java/org/cryptomator/ui/health/CheckExecutor.java

```diff
@@ -0,0 +1,113 @@
+package org.cryptomator.ui.health;
+
+import org.cryptomator.common.vaults.Vault;
+import org.cryptomator.cryptofs.VaultConfig;
+import org.cryptomator.cryptofs.health.api.DiagnosticResult;
+import org.cryptomator.cryptolib.api.CryptorProvider;
+import org.cryptomator.cryptolib.api.Masterkey;
+
+import javax.inject.Inject;
+import javafx.concurrent.Task;
+import java.nio.file.Path;
+import java.security.SecureRandom;
+import java.util.ArrayDeque;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicReference;
+
+@HealthCheckScoped
+public class CheckExecutor {
+
+	private final Path vaultPath;
+	private final SecureRandom csprng;
+	private final Masterkey masterkey;
+	private final VaultConfig vaultConfig;
+	private final ExecutorService sequentialExecutor;
+	private final Queue<CheckTask> runningTasks;
+
+
+	@Inject
+	public CheckExecutor(@HealthCheckWindow Vault vault, AtomicReference<Masterkey> masterkeyRef, AtomicReference<VaultConfig> vaultConfigRef, SecureRandom csprng) {
+		this.vaultPath = vault.getPath();
+		this.masterkey = masterkeyRef.get();
+		this.vaultConfig = vaultConfigRef.get();
+		this.csprng = csprng;
+		this.sequentialExecutor = Executors.newSingleThreadExecutor();
+		this.runningTasks = new ArrayDeque<>();
```

Comment:
If we use `new ThreadPoolExecutor(...)`, we can specify the task queue as a param and use it later. This avoids having two queues holding tasks...

Reply:
A good idea, but there is an edge case which prevents its usage: The queue is used only for scheduled tasks, an already running task cannot be canceled anymore except by shutting down the executor.

Reply:
What a shame